addons/website_slides/models/slide_slide.py

```diff
@@ -335,58 +361,116 @@ def _compute_user_membership_id(self):
+                if slide.video_source_type == 'youtube':
+                    query_params = urls.url_parse(slide.video_url).query
+                    query_params = query_params + '&theme=light' if query_params else 'theme=light'
+                    embed_code = '<iframe src="//www.youtube-nocookie.com/embed/%s?%s" allowFullScreen="true" frameborder="0"></iframe>' % (slide.video_youtube_id, query_params)
+                elif slide.video_source_type == 'google_drive':
+                    embed_code = '<iframe src="//drive.google.com/file/d/%s/preview" allowFullScreen="true" frameborder="0"></iframe>' % (slide.video_google_drive_id)
+            elif slide.slide_type in ['infographic', 'document'] and slide.source_type == 'external' and slide.document_google_drive_id:
+                embed_code = '<iframe src="//drive.google.com/file/d/%s/preview" allowFullScreen="true" frameborder="0"></iframe>' % (slide.document_google_drive_id)
+            elif slide.slide_type == 'document' and slide.source_type == 'local_file':
+                slide_url = base_url + url_for('/slides/embed/%s?page=1' % slide.id)
+                embed_code = '<iframe src="%s" class="o_wslides_iframe_viewer" allowFullScreen="true" height="%s" width="%s" frameborder="0"></iframe>' % (slide_url, 315, 420)
+
+            slide.embed_code = embed_code
+
+    @api.depends('video_url')
+    def _compute_video_source_type(self):
+        for slide in self:
+            video_source_type = False
+            youtube_match = re.match(self.YOUTUBE_VIDEO_ID_REGEX, slide.video_url) if slide.video_url else False
+            if youtube_match:
+                if youtube_match and len(youtube_match.groups()) == 2 and len(youtube_match.group(2)) == 11:
+                    video_source_type = 'youtube'
+            if slide.video_url and not video_source_type and re.match(self.GOOGLE_DRIVE_DOCUMENT_ID_REGEX, slide.video_url):
+                video_source_type = 'google_drive'
+
+            slide.video_source_type = video_source_type
+
+    @api.depends('video_url', 'video_source_type')
+    def _compute_video_youtube_id(self):
+        for slide in self:
+            if slide.video_url and slide.video_source_type == 'youtube':
+                match = re.match(self.YOUTUBE_VIDEO_ID_REGEX, slide.video_url)
+                if match and len(match.groups()) == 2 and len(match.group(2)) == 11:
+                    slide.video_youtube_id = match.group(2)
                 else:
-                    # embed google doc video
-                    record.embed_code = '<iframe src="//drive.google.com/file/d/%s/preview" allowFullScreen="true" frameborder="0"></iframe>' % (record.document_id)
+                    slide.video_youtube_id = False
             else:
-                record.embed_code = False
+                slide.video_youtube_id = False
 
-    @api.onchange('url')
+    @api.depends('slide_type', 'document_url', 'video_url', 'video_source_type')
```

Comment:
Somehow I think having a unique stored url field could help as it would depend on external_url only. Would also allow to remove dependency on video_source_type itself, as in the end it is not really important for google drive id computation if I understand well.